smqtk_core/__init__.py

```diff
@@ -1,12 +1,64 @@
+from __future__ import annotations
+
 import abc
+from collections.abc import Mapping
+from typing import TYPE_CHECKING, Any
 
-from .configuration import Configurable  # noqa: F401
+from .configuration import Configurable, from_config_dict  # noqa: F401
 from .plugin import Pluggable  # noqa: F401
 
 
-class Plugfigurable (Pluggable, Configurable, metaclass=abc.ABCMeta):
+if TYPE_CHECKING:
+    import pydantic
+    from pydantic_core import core_schema
+
+
+class Plugfigurable(Pluggable, Configurable, metaclass=abc.ABCMeta):
     """
     When you don't want to have to constantly inherit from two mixin classes
     all the time, we provide this as a convenience that descends from both
     mixin classes: Pluggable and Configurable.
     """
+
+    try:
+        import pydantic
+        from pydantic_core import core_schema
+
+        PYDANTIC_2_AVAILABLE = tuple(map(int, pydantic.__version__.split(".")[:3])) >= (2,)
+    except (ImportError, AttributeError):
+        PYDANTIC_2_AVAILABLE = False
+
+    if PYDANTIC_2_AVAILABLE:
+        @classmethod
+        def __get_pydantic_core_schema__(
+            cls, _source_type: Any, _handler: pydantic.GetCoreSchemaHandler
+        ) -> core_schema.CoreSchema:
+            from pydantic_core import core_schema
+
+            def from_config(value: Mapping[str, Any]) -> Plugfigurable:
+                return from_config_dict(dict(value), cls.get_impls())
```

Comment:
I'm getting an error when running your example where this fails with the error `TypeError: 'module' object is not callable`. It seems like when this context is run `dict` is referring to the `smqtk_core.dict` submodule. I'm not totally sure how this is happening, but I would not be surprised it the not-great-naming if the `dict.py` utility submodule in combination with `Plugfigurable` being defined in this `__init__` module is seeming to rear its head. Given your suggestion of moving `Plugfigurable` into a separated submodule (i.e. `smqtk_core/plugfigurable.py`) is one that I agree with an would at least locally address this conflict.